// @ui5/webcomponents-react/ssr/components/FormGroup/index.js
import React from 'react';
import { createUseStyles } from 'react-jss';
import { FormGroupTitle } from './FormGroupTitle';
import { jsx as _jsx } from "react/jsx-runtime";
import { Fragment as _Fragment } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
const useStyles = createUseStyles({
  spacer: {
    height: '1rem',
    gridColumn: 'span 12'
  }
}, {
  name: 'FormGroup'
});
/**
 * The `FormGroup` encapsulates `FormItems` into groups.
 *
 * __Note:__ `FormGroup` is only used for calculating the final layout of the `Form`, thus it doesn't accept any other props than `titleText` and `children`, especially no `className`, `style` or `ref`.
 */
const FormGroup = props => {
  const {
    titleText,
    children
  } = props;
  const classes = useStyles();
  return /*#__PURE__*/_jsxs(_Fragment, {
    children: [/*#__PURE__*/_jsx(FormGroupTitle, {
      titleText: titleText
    }), children, /*#__PURE__*/_jsx("span", {
      className: classes.spacer
    })]
  });
};
FormGroup.displayName = 'FormGroup';
export { FormGroup };
