// @ui5/webcomponents-react/ssr/components/ObjectPageSection/index.js
'use client';

import { clsx } from 'clsx';
import React, { forwardRef } from 'react';
import { createUseStyles } from 'react-jss';
import styles from './ObjectPageSection.jss';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
const useStyles = createUseStyles(styles, {
  name: 'ObjectPageSection'
});
/**
 * Top-level information container of an `ObjectPage`.
 */
const ObjectPageSection = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    titleText = '',
    id,
    children,
    titleTextUppercase = true,
    className,
    style,
    hideTitleText,
    titleTextLevel = 'H3',
    wrapTitleText,
    header,
    ...rest
  } = props;
  const classes = useStyles();
  const htmlId = `ObjectPageSection-${id}`;
  const titleClasses = clsx(classes.title, titleTextUppercase && classes.uppercase);
  return /*#__PURE__*/_jsxs("section", {
    ref: ref,
    role: "region",
    className: clsx(classes.section, wrapTitleText && classes.wrap, className),
    style: style,
    ...rest,
    id: htmlId,
    "data-component-name": "ObjectPageSection",
    children: [!!header && /*#__PURE__*/_jsx("div", {
      className: classes.headerContainer,
      children: header
    }), !hideTitleText && /*#__PURE__*/_jsx("div", {
      role: "heading",
      "aria-level": parseInt(titleTextLevel.slice(1)),
      className: classes.titleContainer,
      "data-component-name": "ObjectPageSectionTitleText",
      children: /*#__PURE__*/_jsx("div", {
        className: titleClasses,
        children: titleText
      })
    }), /*#__PURE__*/_jsx("div", {
      className: classes.sectionContent,
      children: /*#__PURE__*/_jsx("div", {
        className: classes.sectionContentInner,
        "data-component-name": "ObjectPageSectionContent",
        children: children
      })
    })]
  });
});
ObjectPageSection.displayName = 'ObjectPageSection';
export { ObjectPageSection };
