// @ui5/webcomponents-react/dist/components/FilterBar/index.js
'use client';

import { debounce, Device, enrichEventWithDetails, useI18nBundle } from '@ui5/webcomponents-react-base';
import { clsx } from 'clsx';
import React, { Children, cloneElement, forwardRef, isValidElement, useEffect, useRef, useState } from 'react';
import { createUseStyles } from 'react-jss';
import { ButtonDesign, ToolbarStyle } from '../../enums';
import { ADAPT_FILTERS, CLEAR, FILTERS, GO, HIDE_FILTER_BAR, RESTORE, SHOW_FILTER_BAR } from '../../i18n/i18n-defaults';
import { Button } from '../../webComponents';
import { Toolbar } from '../Toolbar';
import { ToolbarSeparator } from '../ToolbarSeparator';
import { ToolbarSpacer } from '../ToolbarSpacer';
import styles from './FilterBar.jss';
import { FilterDialog } from './FilterDialog';
import { filterValue, renderSearchWithValue, syncRef } from './utils';
const isPhone = Device.isPhone();
const isTablet = Device.isTablet();
const resizeObserverEntryWidth = entry => {
  if (entry.borderBoxSize) {
    // Firefox implements `borderBoxSize` as a single content rect, rather than an array
    return Array.isArray(entry.borderBoxSize) ? entry.borderBoxSize[0]?.inlineSize : entry.borderBoxSize?.inlineSize;
  }
  // Safari doesn't implement `borderBoxSize`
  return entry.target.getBoundingClientRect().width;
};
const useStyles = createUseStyles(styles, {
  name: 'FilterBar'
});
/**
 * The `FilterBar` displays filters in a user-friendly manner to populate values for a query. It consists of a row containing the `VariantManagement` or a title, the related buttons, and an area underneath displaying the filters. The filters are arranged in a logical row that is divided depending on the space available and the width of the filters. The area containing the filters can be hidden or shown using the "Hide FilterBar / Show FilterBar" button, the "Filters" button shows the filter dialog.
 In this dialog, the consumer has full control over the FilterBar. The filters in this dialog are displayed in one column and organized in groups. Each filter can be marked as visible in the FilterBar by selecting "Add to FilterBar".
 */
const FilterBar = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    children,
    hideToolbar,
    filterBarCollapsed,
    considerGroupName,
    filterContainerWidth = '13.125rem',
    activeFiltersCount,
    showClearOnFB,
    showGoOnFB,
    hideFilterConfiguration,
    showRestoreOnFB,
    showResetButton,
    hideToggleFiltersButton,
    style,
    className,
    slot,
    search,
    header,
    as = 'div',
    portalContainer,
    onToggleFilters,
    onFiltersDialogOpen,
    onAfterFiltersDialogOpen,
    onFiltersDialogCancel,
    onFiltersDialogClose,
    onFiltersDialogSave,
    onClear,
    onFiltersDialogSelectionChange,
    onFiltersDialogSearch,
    onGo,
    onRestore,
    ...rest
  } = props;
  const initiallyShowFilters = (() => {
    if (!hideToolbar) {
      if (filterBarCollapsed !== undefined) {
        return filterBarCollapsed;
      }
      return !isTablet;
    }
    return true;
  })();
  const [showFilters, setShowFilters] = useState(initiallyShowFilters);
  const [mountFilters, setMountFilters] = useState(true);
  const [dialogOpen, setDialogOpen] = useState(false);
  const [searchValue, setSearchValue] = useState(undefined);
  const searchRef = useRef(null);
  const filterRefs = useRef({});
  const dialogRef = useRef(null);
  const [dialogRefs, setDialogRefs] = useState({});
  const [toggledFilters, setToggledFilters] = useState({});
  const prevVisibleInFilterBarProps = useRef({});
  const prevSearchInputPropsValueRef = useRef();
  const filterBarButtonsRef = useRef(null);
  const filterAreaRef = useRef(null);
  const filterBtnRef = useRef(null);
  const i18nBundle = useI18nBundle('@ui5/webcomponents-react');
  const clearText = i18nBundle.getText(CLEAR);
  const restoreText = i18nBundle.getText(RESTORE);
  const showFilterBarText = i18nBundle.getText(SHOW_FILTER_BAR);
  const hideFilterBarText = i18nBundle.getText(HIDE_FILTER_BAR);
  const goText = i18nBundle.getText(GO);
  const filtersText = !hideToolbar ? i18nBundle.getText(FILTERS) : i18nBundle.getText(ADAPT_FILTERS);

  // dialog
  const [isListView, setIsListView] = useState(true);
  const [filteredAttribute, setFilteredAttribute] = useState('all');
  useEffect(() => {
    Children.toArray(children).forEach(item => {
      if ( /*#__PURE__*/isValidElement(item)) {
        setToggledFilters(prev => {
          if (!item.props.hasOwnProperty('visibleInFilterBar') && prev?.[item.key] === undefined) {
            return {
              ...prev,
              [item.key]: true
            };
          }
          if (item.props.hasOwnProperty('visibleInFilterBar')) {
            return {
              ...prev,
              [item.key]: item.props.visibleInFilterBar
            };
          }
          return prev;
        });
      }
    });
  }, [children, setToggledFilters]);
  useEffect(() => {
    if (filterBarCollapsed !== undefined) {
      setShowFilters(!hideToolbar ? !filterBarCollapsed : true);
    }
  }, [setShowFilters, hideToolbar, filterBarCollapsed]);
  const classes = useStyles();
  const filterAreaClasses = clsx(classes.filterArea, showFilters && (!isPhone || isPhone && hideToolbar) ? classes.filterAreaOpen : classes.filterAreaClosed);
  const getFilterElements = () => {
    const search = searchRef.current?.querySelector(`[data-component-name="FilterBarSearch"]`);
    return {
      filters: filterRefs?.current ? Object.values(filterRefs.current).filter(Boolean) : [],
      search
    };
  };
  const handleToggle = e => {
    if (onToggleFilters) {
      onToggleFilters(enrichEventWithDetails(e, {
        visible: !showFilters,
        ...getFilterElements()
      }));
    }
    setShowFilters(!showFilters);
  };
  const [executeGo, setExecuteGo] = useState(false);
  const handleDialogSave = (e, newRefs, updatedToggledFilters) => {
    setDialogRefs(newRefs);
    const details = {
      elements: newRefs,
      toggledElements: {
        ...toggledFilters,
        ...updatedToggledFilters
      },
      ...getFilterElements()
    };
    setToggledFilters(old => ({
      ...old,
      ...updatedToggledFilters
    }));
    if (onFiltersDialogSave) {
      onFiltersDialogSave(enrichEventWithDetails(e, details));
    }
    handleDialogClose(e);
  };
  const handleDialogOpen = e => {
    if (typeof onFiltersDialogOpen === 'function') {
      onFiltersDialogOpen(e);
    }
    if (e.defaultPrevented) {
      setDialogOpen(false);
    } else {
      setDialogOpen(true);
    }
  };
  const handleDialogClose = e => {
    if (onFiltersDialogClose) {
      onFiltersDialogClose(enrichEventWithDetails(e));
    }
    setDialogOpen(false);
    filterBtnRef.current?.focus();
  };
  const handleGoOnFb = e => {
    if (typeof onGo === 'function') {
      onGo(enrichEventWithDetails(e, {
        elements: filterRefs.current,
        ...getFilterElements()
      }));
    }
  };
  const safeChildren = () => {
    if (Object.keys(toggledFilters).length > 0) {
      return Children.toArray(children).map(child => {
        if ( /*#__PURE__*/isValidElement(child) && toggledFilters?.[child.key] !== undefined) {
          // @ts-expect-error: child should always be a FilterGroupItem w/o portal
          return /*#__PURE__*/cloneElement(child, {
            visibleInFilterBar: toggledFilters[child.key]
          });
        }
        return child;
      });
    }
    return Children.toArray(children);
  };
  const prevChildren = useRef({});
  const renderChildren = () => {
    const childProps = {
      considerGroupName,
      ['data-in-fb']: true,
      ['data-with-toolbar']: !hideToolbar
    };
    return safeChildren().filter(item => {
      if (! /*#__PURE__*/isValidElement(item)) {
        return false;
      }
      return item?.props?.visible && item.props?.visibleInFilterBar;
    }).map(child => {
      // necessary because of varying widths of input elements
      if (filterContainerWidth) {
        childProps.style = {
          width: filterContainerWidth,
          ...child.props.style
        };
      }
      if (hideFilterConfiguration) {
        return /*#__PURE__*/cloneElement(child, {
          ...childProps
        });
      }
      prevVisibleInFilterBarProps.current[child.key] = child.props.visibleInFilterBar;
      let filterItemProps = {};
      if (Object.keys(dialogRefs).length > 0) {
        const dialogItemRef = dialogRefs[child.key];
        if (dialogItemRef) {
          filterItemProps = filterValue(dialogItemRef, child);
        }
      }
      if (!child.props.children) {
        return /*#__PURE__*/cloneElement(child, {
          ...childProps
        });
      }
      if (prevChildren.current?.[child.key] && (
      //Input
      child.props.children?.props?.value !== prevChildren.current?.[child.key]?.value ||
      //Checkbox
      child.props.children?.props?.checked !== prevChildren.current?.[child.key]?.checked ||
      //Selectable
      Array.isArray(child.props.children?.props?.children) && child.props.children?.props?.children?.map(item => item.props.selected).join(',') !== prevChildren?.current?.[child.key]?.children?.map(item => item.props.selected).join(','))) {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        const {
          [child.key]: _omit,
          ...rest
        } = dialogRefs;
        setDialogRefs(rest);
      }
      prevChildren.current[child.key] = child.props.children.props;
      return /*#__PURE__*/cloneElement(child, {
        ...childProps,
        children: {
          ...child.props.children,
          props: {
            ...child.props.children.props,
            ...filterItemProps
          },
          ref: node => {
            filterRefs.current[child.key] = node;
            if (!dialogOpen) syncRef(child.props.children.ref, node);
          }
        }
      });
    });
  };
  const handleRestoreFilters = (e, source, filterElements) => {
    if (source === 'filterBar' && showGoOnFB) {
      setMountFilters(false);
    }
    if (onRestore) {
      onRestore(enrichEventWithDetails(e, {
        source,
        ...filterElements
      }));
    }
  };
  useEffect(() => {
    if (!mountFilters) {
      setMountFilters(true);
    }
  }, [mountFilters]);
  const handleFBRestore = e => {
    handleRestoreFilters(e, 'filterBar', getFilterElements());
  };
  const handleClear = e => {
    if (typeof onClear === 'function') {
      onClear(enrichEventWithDetails(e, getFilterElements()));
    }
  };
  const cssClasses = clsx(classes.outerContainer, className, !hideToolbar && classes.outerContainerWithToolbar);
  useEffect(() => {
    prevSearchInputPropsValueRef.current = search?.props?.value;
  }, [search?.props?.value]);
  const prevSearchInputPropsValue = prevSearchInputPropsValueRef.current;
  useEffect(() => {
    if (prevSearchInputPropsValue !== search?.props?.value) {
      setSearchValue(search?.props?.value);
    }
  }, [prevSearchInputPropsValue, search?.props?.value]);
  const ToolbarButtons = /*#__PURE__*/React.createElement(React.Fragment, null, showGoOnFB && /*#__PURE__*/React.createElement(Button, {
    onClick: handleGoOnFb,
    design: ButtonDesign.Emphasized
  }, goText), !hideToggleFiltersButton && !hideToolbar && !isPhone && /*#__PURE__*/React.createElement(Button, {
    onClick: handleToggle,
    design: ButtonDesign.Transparent,
    className: classes.showFiltersBtn,
    "aria-live": "polite"
  }, showFilters ? hideFilterBarText : showFilterBarText), showClearOnFB && /*#__PURE__*/React.createElement(Button, {
    onClick: handleClear,
    design: ButtonDesign.Transparent
  }, clearText), showRestoreOnFB && /*#__PURE__*/React.createElement(Button, {
    onClick: handleFBRestore,
    design: ButtonDesign.Transparent
  }, restoreText), !hideFilterConfiguration && /*#__PURE__*/React.createElement(Button, {
    onClick: handleDialogOpen,
    "aria-haspopup": "dialog",
    design: ButtonDesign.Transparent,
    ref: filterBtnRef
  }, `${filtersText}${activeFiltersCount && parseInt(activeFiltersCount, 10) > 0 ? ` (${activeFiltersCount})` : ''}`));
  const hasButtons = ToolbarButtons.props.children.some(Boolean);
  const [filterBarButtonsWidth, setFilterBarButtonsWidth] = useState(undefined);
  const [filterAreaWidth, setFilterAreaWidth] = useState(undefined);
  const [firstChildWidth, setFirstChildWidth] = useState(undefined);
  useEffect(() => {
    const debouncedObserverFn = debounce(_ref => {
      let [area] = _ref;
      const firstChild = area.target?.children?.[0];
      if (firstChild && firstChild.offsetWidth !== firstChildWidth) {
        setFirstChildWidth(firstChild.offsetWidth + 16 /*margin*/);
      }
    }, 100);
    const filterAreaObserver = new ResizeObserver(debouncedObserverFn);
    if (hideToolbar && filterAreaRef.current) {
      filterAreaObserver.observe(filterAreaRef.current);
    }
    return () => {
      debouncedObserverFn.cancel();
      filterAreaObserver.disconnect();
    };
  }, [filterAreaRef.current, hideToolbar]);
  useEffect(() => {
    const debouncedObserverFn = debounce(_ref2 => {
      let [area] = _ref2;
      const filterWidth = resizeObserverEntryWidth(area);
      if (filterWidth !== filterBarButtonsWidth) {
        setFilterAreaWidth(filterWidth);
      }
    }, 100);
    const filterAreaObserver = new ResizeObserver(debouncedObserverFn);
    if (hideToolbar && filterAreaRef.current) {
      filterAreaObserver.observe(filterAreaRef.current);
    }
    return () => {
      debouncedObserverFn.cancel();
      filterAreaObserver.disconnect();
    };
  }, [filterAreaWidth, filterAreaRef.current, hideToolbar]);
  useEffect(() => {
    const debouncedObserverFn = debounce(_ref3 => {
      let [buttons] = _ref3;
      const buttonsWidth = resizeObserverEntryWidth(buttons);
      if (buttonsWidth !== filterBarButtonsWidth) {
        setFilterBarButtonsWidth(buttonsWidth);
      }
    }, 100);
    const filterBarButtonsObserver = new ResizeObserver(debouncedObserverFn);
    if (hideToolbar && filterBarButtonsRef.current) {
      filterBarButtonsObserver.observe(filterBarButtonsRef.current);
    }
    return () => {
      debouncedObserverFn.cancel();
      filterBarButtonsObserver.disconnect();
    };
  }, [filterBarButtonsRef.current, hideToolbar, filterBarButtonsWidth]);
  const calculatedChildren = renderChildren();
  useEffect(() => {
    if (executeGo) {
      setExecuteGo(prevEvent => {
        onGo(prevEvent);
        return false;
      });
    }
  }, [executeGo]);

  // calculates the number of spacers depending on the available width inside the row
  const renderSpacers = () => {
    if (firstChildWidth && filterAreaWidth && filterBarButtonsWidth) {
      const spacers = [];
      const filterItemsWidth = calculatedChildren.length * firstChildWidth;
      //early return if enough space is available
      if (filterAreaWidth - filterBarButtonsWidth > filterItemsWidth) {
        return null;
      }
      const usedSpaceLastRow = filterItemsWidth % filterAreaWidth;
      const emptySpaceLastRow = filterAreaWidth - usedSpaceLastRow;
      // deduct width of buttons container of the empty space in the last row to calculate number of spacers
      const numberOfSpacers = Math.floor((emptySpaceLastRow - filterBarButtonsWidth) / firstChildWidth);
      for (let i = 0; i < numberOfSpacers; i++) {
        spacers.push( /*#__PURE__*/React.createElement("div", {
          key: `filter-spacer-${i}`,
          className: classes.spacer
        }));
      }
      return spacers;
    }
    return null;
  };
  const CustomTag = as;
  return /*#__PURE__*/React.createElement(React.Fragment, null, dialogOpen && !hideFilterConfiguration && /*#__PURE__*/React.createElement(FilterDialog, {
    filterBarRefs: filterRefs,
    open: dialogOpen,
    handleDialogClose: handleDialogClose,
    handleRestoreFilters: handleRestoreFilters,
    handleSearchValueChange: setSearchValue,
    showRestoreButton: showResetButton,
    handleSelectionChange: onFiltersDialogSelectionChange,
    handleDialogSave: handleDialogSave,
    handleDialogSearch: onFiltersDialogSearch,
    handleDialogCancel: onFiltersDialogCancel,
    onAfterFiltersDialogOpen: onAfterFiltersDialogOpen,
    portalContainer: portalContainer,
    dialogRef: dialogRef,
    isListView: isListView,
    setIsListView: setIsListView,
    filteredAttribute: filteredAttribute,
    setFilteredAttribute: setFilteredAttribute
  }, safeChildren()), /*#__PURE__*/React.createElement(CustomTag, {
    ref: ref,
    className: cssClasses,
    style: {
      ['--_ui5wcr_filter_group_item_flex_basis']: filterContainerWidth,
      ...style
    },
    slot: slot,
    ...rest
  }, !hideToolbar && /*#__PURE__*/React.createElement(Toolbar, {
    className: classes.filterBarHeader,
    toolbarStyle: ToolbarStyle.Clear
  }, header, header && search && /*#__PURE__*/React.createElement(ToolbarSeparator, null), search && !isPhone && /*#__PURE__*/React.createElement("div", {
    ref: searchRef
  }, renderSearchWithValue(search, searchValue)), hasButtons && /*#__PURE__*/React.createElement(ToolbarSpacer, null), ToolbarButtons), mountFilters && /*#__PURE__*/React.createElement("div", {
    className: filterAreaClasses,
    style: {
      position: 'relative'
    },
    ref: filterAreaRef
  }, calculatedChildren, hideToolbar && /*#__PURE__*/React.createElement(React.Fragment, null, renderSpacers(), /*#__PURE__*/React.createElement("div", {
    style: {
      width: filterBarButtonsWidth ? `${filterBarButtonsWidth}px` : '120px',
      minWidth: filterBarButtonsWidth ? `${filterBarButtonsWidth}px` : '120px'
    },
    className: classes.lastSpacer
  }, /*#__PURE__*/React.createElement("div", {
    className: classes.filterBarButtons,
    ref: filterBarButtonsRef
  }, ToolbarButtons))))));
});
FilterBar.displayName = 'FilterBar';
export { FilterBar };
