// @ui5/webcomponents-react/ssr/components/AnalyticalTable/ColumnHeader/ColumnHeaderModal.js
import iconDecline from '@ui5/webcomponents-icons/dist/decline.js';
import iconFilter from '@ui5/webcomponents-icons/dist/filter.js';
import iconGroup from '@ui5/webcomponents-icons/dist/group-2.js';
import iconSortAscending from '@ui5/webcomponents-icons/dist/sort-ascending.js';
import iconSortDescending from '@ui5/webcomponents-icons/dist/sort-descending.js';
import { enrichEventWithDetails, ThemingParameters, useI18nBundle } from '@ui5/webcomponents-react-base';
import React, { useEffect, useRef } from 'react';
import { createPortal } from 'react-dom';
import { createUseStyles } from 'react-jss';
import { FlexBoxAlignItems } from '../../../enums/FlexBoxAlignItems';
import { ListItemType } from '../../../enums/ListItemType';
import { PopoverHorizontalAlign } from '../../../enums/PopoverHorizontalAlign';
import { PopoverPlacementType } from '../../../enums/PopoverPlacementType';
import { TextAlign } from '../../../enums/TextAlign';
import { CLEAR_SORTING, GROUP, SORT_ASCENDING, SORT_DESCENDING, UNGROUP } from '../../../i18n/i18n-defaults';
import { useCanRenderPortal } from '../../../internal/ssr';
import { stopPropagation } from '../../../internal/stopPropagation';
import { getUi5TagWithSuffix } from '../../../internal/utils';
import { CustomListItem } from '../../../webComponents/CustomListItem';
import { Icon } from '../../../webComponents/Icon';
import { List } from '../../../webComponents/List';
import { Popover } from '../../../webComponents/Popover';
import { StandardListItem } from '../../../webComponents/StandardListItem';
import { FlexBox } from '../../FlexBox';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
const styles = {
  popover: {
    fontWeight: 'normal',
    '&::part(content)': {
      padding: 0
    }
  },
  filterListItem: {
    height: 'var(--_ui5_list_item_base_height)'
  },
  filter: {
    height: 'var(--_ui5_list_item_base_height)'
  },
  filterIcon: {
    paddingInlineEnd: '0.5rem',
    minWidth: '1rem',
    minHeight: '1rem',
    color: ThemingParameters.sapContent_NonInteractiveIconColor
  }
};
const useStyles = createUseStyles(styles, {
  name: 'ColumnHeaderModal'
});
export const ColumnHeaderModal = props => {
  const {
    column,
    onSort,
    onGroupBy,
    open,
    setPopoverOpen,
    portalContainer,
    isRtl,
    openerRef
  } = props;
  const classes = useStyles();
  const showFilter = column.canFilter;
  const showGroup = column.canGroupBy;
  const showSort = column.canSort;
  const ref = useRef(null);
  const listRef = useRef(null);
  const {
    Filter
  } = column;
  const i18nBundle = useI18nBundle('@ui5/webcomponents-react');
  const clearSortingText = i18nBundle.getText(CLEAR_SORTING);
  const sortAscendingText = i18nBundle.getText(SORT_ASCENDING);
  const sortDescendingText = i18nBundle.getText(SORT_DESCENDING);
  const groupText = i18nBundle.getText(GROUP);
  const ungroupText = i18nBundle.getText(UNGROUP);
  const handleSort = e => {
    const sortType = e.detail.item.getAttribute('data-sort');
    switch (sortType) {
      case 'asc':
        column.toggleSortBy(false);
        if (typeof onSort === 'function') {
          onSort(enrichEventWithDetails(e, {
            column,
            sortDirection: sortType
          }));
        }
        break;
      case 'desc':
        column.toggleSortBy(true);
        if (typeof onSort === 'function') {
          onSort(enrichEventWithDetails(e, {
            column,
            sortDirection: sortType
          }));
        }
        break;
      case 'clear':
        column.clearSortBy();
        if (typeof onSort === 'function') {
          onSort(enrichEventWithDetails(e, {
            column,
            sortDirection: sortType
          }));
        }
        break;
      case 'group':
        const willGroup = !column.isGrouped;
        column.toggleGroupBy(willGroup);
        if (typeof onGroupBy === 'function') {
          onGroupBy(enrichEventWithDetails(e, {
            column,
            isGrouped: willGroup
          }));
        }
        break;
    }
    setPopoverOpen(false);
  };
  const isSortedAscending = column.isSorted && column.isSortedDesc === false;
  const isSortedDescending = column.isSorted && column.isSortedDesc === true;
  const onAfterClose = e => {
    stopPropagation(e);
    setPopoverOpen(false);
  };
  const onAfterOpen = () => {
    listRef.current?.children?.[0]?.focus();
  };
  const horizontalAlign = (() => {
    switch (column.hAlign) {
      case TextAlign.Begin:
        return isRtl ? PopoverHorizontalAlign.Right : PopoverHorizontalAlign.Left;
      case TextAlign.End:
        return isRtl ? PopoverHorizontalAlign.Left : PopoverHorizontalAlign.Right;
      case TextAlign.Left:
        return PopoverHorizontalAlign.Left;
      case TextAlign.Right:
        return PopoverHorizontalAlign.Right;
      case TextAlign.Center:
        return PopoverHorizontalAlign.Center;
      default:
        return isRtl ? PopoverHorizontalAlign.Right : PopoverHorizontalAlign.Left;
    }
  })();
  const handleCustomLiKeyDown = e => {
    if (e.key === 'Enter') {
      setPopoverOpen(false);
    }
  };
  const handleListKeyDown = e => {
    if (e.key !== 'Escape') {
      stopPropagation(e);
    }
  };
  const canRenderPortal = useCanRenderPortal();
  useEffect(() => {
    if (open && ref.current && openerRef.current) {
      customElements.whenDefined(getUi5TagWithSuffix('ui5-popover')).then(() => {
        ref.current.opener = openerRef.current;
        if (canRenderPortal && open) {
          ref.current.showAt(openerRef.current);
        }
      }).catch(() => {
        // silently catch
      });
    }
  }, [open, canRenderPortal]);
  if (!canRenderPortal) {
    return null;
  }
  return /*#__PURE__*/createPortal( /*#__PURE__*/_jsx(Popover, {
    hideArrow: true,
    horizontalAlign: horizontalAlign,
    placementType: PopoverPlacementType.Bottom,
    ref: ref,
    className: classes.popover,
    onClick: stopPropagation,
    onAfterClose: onAfterClose,
    onAfterOpen: onAfterOpen,
    children: /*#__PURE__*/_jsxs(List, {
      onItemClick: handleSort,
      ref: listRef,
      onKeyDown: handleListKeyDown,
      children: [isSortedAscending && /*#__PURE__*/_jsx(StandardListItem, {
        type: ListItemType.Active,
        icon: iconDecline,
        "data-sort": "clear",
        children: clearSortingText
      }), showSort && !isSortedAscending && /*#__PURE__*/_jsx(StandardListItem, {
        type: ListItemType.Active,
        icon: iconSortAscending,
        "data-sort": "asc",
        children: sortAscendingText
      }), showSort && !isSortedDescending && /*#__PURE__*/_jsx(StandardListItem, {
        type: ListItemType.Active,
        icon: iconSortDescending,
        "data-sort": "desc",
        children: sortDescendingText
      }), isSortedDescending && /*#__PURE__*/_jsx(StandardListItem, {
        type: ListItemType.Active,
        icon: iconDecline,
        "data-sort": "clear",
        children: clearSortingText
      }), showFilter && !column.isGrouped &&
      /*#__PURE__*/
      //todo maybe need to enhance Input selection after ui5-webcomponents issue has been fixed (undefined is displayed as val)
      _jsx(CustomListItem, {
        type: ListItemType.Inactive,
        className: classes.filterListItem,
        onKeyDown: handleCustomLiKeyDown,
        children: /*#__PURE__*/_jsxs(FlexBox, {
          alignItems: FlexBoxAlignItems.Center,
          className: classes.filter,
          children: [/*#__PURE__*/_jsx(Icon, {
            name: iconFilter,
            className: classes.filterIcon,
            "aria-hidden": true
          }), /*#__PURE__*/_jsx(Filter, {
            column: column,
            popoverRef: ref
          })]
        })
      }), showGroup && /*#__PURE__*/_jsx(StandardListItem, {
        type: ListItemType.Active,
        icon: iconGroup,
        "data-sort": 'group',
        children: column.isGrouped ? ungroupText : groupText
      })]
    })
  }), portalContainer ?? document.body);
};
ColumnHeaderModal.displayName = 'ColumnHeaderModal';
