// @ui5/webcomponents-react/ssr/components/AnalyticalTable/hooks/useRowNavigationIndicator.js
import React from 'react';
import { jsx as _jsx } from "react/jsx-runtime";
const baseStyles = {
  width: '100%',
  height: '100%'
};

/*
 * COMPONENTS
 */
const Header = () => /*#__PURE__*/_jsx("div", {
  style: {
    width: '6px'
  }
});
const Cell = instance => {
  const {
    isNavigatedCell,
    webComponentsReactProperties
  } = instance;
  if (isNavigatedCell) {
    return /*#__PURE__*/_jsx("div", {
      style: baseStyles,
      className: webComponentsReactProperties.classes.valueStateInformation,
      "data-component-name": "AnalyticalTableNavigatedCell"
    });
  }
  return /*#__PURE__*/_jsx("div", {
    style: baseStyles
  });
};

/*
 * TABLE HOOKS
 */
const columnsDeps = (deps, _ref) => {
  let {
    instance: {
      webComponentsReactProperties
    }
  } = _ref;
  return [...deps, webComponentsReactProperties.withNavigationHighlight];
};
const visibleColumnsDeps = (deps, _ref2) => {
  let {
    instance
  } = _ref2;
  return [...deps, instance.webComponentsReactProperties.withNavigationHighlight];
};
const visibleColumns = (currentVisibleColumns, _ref3) => {
  let {
    instance: {
      webComponentsReactProperties
    }
  } = _ref3;
  if (!webComponentsReactProperties.withNavigationHighlight) {
    return currentVisibleColumns.filter(_ref4 => {
      let {
        id
      } = _ref4;
      return id !== '__ui5wcr__internal_navigation_column';
    });
  }
  const highlightColumn = currentVisibleColumns.find(_ref5 => {
    let {
      id
    } = _ref5;
    return id === '__ui5wcr__internal_navigation_column';
  });
  return [...currentVisibleColumns.filter(_ref6 => {
    let {
      id
    } = _ref6;
    return id !== '__ui5wcr__internal_navigation_column';
  }), highlightColumn];
};
const columns = (currentColumns, _ref7) => {
  let {
    instance
  } = _ref7;
  const {
    withNavigationHighlight
  } = instance.webComponentsReactProperties;
  if (!withNavigationHighlight) {
    return currentColumns;
  }
  return [...currentColumns, {
    id: '__ui5wcr__internal_navigation_column',
    disableFilters: true,
    disableSortBy: true,
    disableGroupBy: true,
    disableResizing: true,
    disableDragAndDrop: true,
    width: 6,
    minWidth: 6,
    maxWidth: 6,
    Header,
    Cell
  }];
};
export const useRowNavigationIndicators = hooks => {
  hooks.columns.push(columns);
  hooks.columnsDeps.push(columnsDeps);
  hooks.visibleColumnsDeps.push(visibleColumnsDeps);
  hooks.visibleColumns.push(visibleColumns);
};
