// @ui5/webcomponents-react/ssr/components/AnalyticalTable/scrollbars/VerticalScrollbar.js
import { ThemingParameters } from '@ui5/webcomponents-react-base';
import { clsx } from 'clsx';
import React, { forwardRef } from 'react';
import { createUseStyles } from 'react-jss';
import { FlexBoxDirection } from '../../../enums/FlexBoxDirection';
import { GlobalStyleClasses } from '../../../enums/GlobalStyleClasses';
import { CustomThemingParameters } from '../../../themes/CustomVariables';
import { FlexBox } from '../../FlexBox';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
const styles = {
  headerSection: {
    boxSizing: 'border-box',
    borderBlockStart: CustomThemingParameters.AnalyticalTableOuterBorderBlock,
    borderInlineEnd: CustomThemingParameters.AnalyticalTableOuterBorderInline,
    borderBlockEnd: `${CustomThemingParameters.AnalyticalTableHeaderBorderWidth} solid ${ThemingParameters.sapList_HeaderBorderColor}`,
    backgroundColor: ThemingParameters.sapList_HeaderBackground,
    borderInlineStart: `1px solid ${ThemingParameters.sapList_BorderColor}`,
    marginInlineStart: '-1px'
  },
  scrollbar: {
    overflowY: 'auto',
    borderInlineEnd: CustomThemingParameters.AnalyticalTableOuterBorderInline,
    borderBlockEnd: `1px solid ${ThemingParameters.sapList_TableFooterBorder}`,
    borderInlineStart: `1px solid ${ThemingParameters.sapList_BorderColor}`,
    marginInlineStart: '-1px'
  },
  bottomSection: {
    flexGrow: 1,
    backgroundColor: ThemingParameters.sapList_FooterBackground,
    boxSizing: 'border-box',
    borderInlineEnd: CustomThemingParameters.AnalyticalTableOuterBorderInline
  }
};
const useStyles = createUseStyles(styles, {
  name: 'VerticalScrollbar'
});
export const VerticalScrollbar = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    internalRowHeight,
    tableRef,
    minRows,
    rows,
    handleVerticalScrollBarScroll,
    popInRowHeight,
    tableBodyHeight
  } = props;
  const classes = useStyles();
  const hasHorizontalScrollbar = tableRef?.current?.offsetWidth !== tableRef?.current?.scrollWidth;
  const horizontalScrollbarSectionStyles = clsx(hasHorizontalScrollbar && classes.bottomSection);
  return /*#__PURE__*/_jsxs(FlexBox, {
    direction: FlexBoxDirection.Column,
    style: {
      position: 'relative'
    },
    "data-component-name": "AnalyticalTableVerticalScrollbarContainer",
    children: [/*#__PURE__*/_jsx("div", {
      style: {
        height: `${internalRowHeight}px`
      },
      className: classes.headerSection
    }), /*#__PURE__*/_jsx("div", {
      ref: ref,
      style: {
        height: tableRef.current ? `${tableBodyHeight}px` : '0'
      },
      onScroll: handleVerticalScrollBarScroll,
      "data-native-scrollbar": props['data-native-scrollbar'],
      className: `${GlobalStyleClasses.sapScrollBar} ${classes.scrollbar}`,
      "data-component-name": "AnalyticalTableVerticalScrollbar",
      children: /*#__PURE__*/_jsx("div", {
        style: {
          height: `${Math.max(minRows, rows.length) * popInRowHeight}px`,
          width: '1px',
          backgroundColor: ThemingParameters.sapList_Background
        }
      })
    }), /*#__PURE__*/_jsx("div", {
      className: horizontalScrollbarSectionStyles
    })]
  });
});
VerticalScrollbar.displayName = 'VerticalScrollbar';
