// @ui5/webcomponents-react/ssr/components/AnalyticalTable/defaults/LoadingComponent/TablePlaceholder.js
import { CssSizeVariables, ThemingParameters } from '@ui5/webcomponents-react-base';
import { clsx } from 'clsx';
import React from 'react';
import { createUseStyles } from 'react-jss';
import { resolveCellAlignment } from '../../util';
import { jsx as _jsx } from "react/jsx-runtime";
const getArrayOfLength = len => Array.from(Array(len).keys());
const useStyles = createUseStyles({
  '@keyframes placeholderShimmer': {
    '0%': {
      backgroundPositionX: '100%'
    },
    '100%': {
      backgroundPositionX: '0'
    }
  },
  animation: {
    animationDuration: '2s',
    animationFillMode: 'forwards',
    animationIterationCount: 'infinite',
    animationName: '$placeholderShimmer',
    animationTimingFunction: 'linear',
    backgroundImage: ThemingParameters.sapContent_Placeholderloading_Gradient,
    backgroundRepeat: 'no-repeat',
    backgroundSize: '1000px 104px'
  },
  row: {
    display: 'flex',
    alignItems: 'center',
    height: CssSizeVariables.sapWcrAnalyticalTableRowHeight
  },
  cellContainer: {
    display: 'flex',
    paddingInline: '0.25rem',
    boxSizing: 'border-box'
  },
  cell: {
    height: ThemingParameters.sapFontSize,
    width: '60%',
    borderRadius: ThemingParameters.sapElement_BorderCornerRadius
  }
}, {
  name: 'AnalyticalTablePlaceholder '
});
export const TablePlaceholder = props => {
  const {
    columns,
    rows,
    style
  } = props;
  const classes = useStyles();
  return /*#__PURE__*/_jsx("div", {
    style: {
      backgroundColor: ThemingParameters.sapList_Background,
      width: '100%',
      ...style
    },
    "data-component-name": "AnalyticalTableLoadingPlaceholder",
    children: getArrayOfLength(rows).map((_, index) => {
      return /*#__PURE__*/_jsx("div", {
        className: classes.row,
        children: columns.map(col => {
          return /*#__PURE__*/_jsx("div", {
            className: classes.cellContainer,
            style: {
              width: col.totalWidth,
              ...resolveCellAlignment(col)
            },
            children: /*#__PURE__*/_jsx("div", {
              className: clsx(classes.cell, classes.animation)
            })
          }, `row${index}-${col.id}`);
        })
      }, `row-${index}`);
    })
  });
};
TablePlaceholder.displayName = 'TablePlaceholder';
TablePlaceholder.defaultProps = {
  rows: 5
};
