// @ui5/webcomponents-react/ssr/webComponents/Carousel/index.js
'use client';

import { CarouselArrowsPlacement, CarouselPageIndicatorStyle } from '../../enums';
import { withWebComponent } from '../../internal/withWebComponent';
/**
 * The Carousel allows the user to browse through a set of items. The component is mostly used for showing a gallery of images, but can hold any other HTML element.
 * There are several ways to perform navigation:
 *
 * *   on desktop - the user can navigate using the navigation arrows or with keyboard shorcuts.
 * *   on mobile - the user can use swipe gestures.
 *
 * __Note:__ This component is a web component developed by the UI5 Web Components’ team.
 *
 * <ui5-link href="https://sap.github.io/ui5-webcomponents/playground/components/Carousel" target="_blank">UI5 Web Components Playground</ui5-link>
 */
const Carousel = withWebComponent('ui5-carousel', ['arrowsPlacement', 'itemsPerPageL', 'itemsPerPageM', 'itemsPerPageS', 'pageIndicatorStyle'], ['cyclic', 'hideNavigationArrows', 'hidePageIndicator'], [], ['navigate'], () => import('@ui5/webcomponents/dist/Carousel.js'));
Carousel.displayName = 'Carousel';
Carousel.defaultProps = {
  arrowsPlacement: CarouselArrowsPlacement.Content,
  itemsPerPageL: 1,
  itemsPerPageM: 1,
  itemsPerPageS: 1,
  pageIndicatorStyle: CarouselPageIndicatorStyle.Default
};
export { Carousel };
