// @ui5/webcomponents-react/ssr/themes/CustomVariables.js
export let CustomVariables = /*#__PURE__*/function (CustomVariables) {
  CustomVariables["ActionSheetBoxShadow"] = "--ui5wcr_ActionSheet_BoxShadow";
  CustomVariables["ActionSheetHeaderBoxShadow"] = "--ui5wcr_ActionSheet_HeaderBoxShadow";
  CustomVariables["ActionSheetMobileHeaderBackground"] = "--ui5wcr_ActionSheet_MobileHeaderBackground";
  CustomVariables["ActionSheetMobileHeaderBorderRadius"] = "--ui5wcr_ActionSheet_MobileHeaderBorderRadius";
  CustomVariables["ActionSheetMobileHeaderBoxShadow"] = "--ui5wcr_ActionSheet_MobileHeaderBoxShadow";
  CustomVariables["ActionSheetMobileHeaderTextColor"] = "--ui5wcr_ActionSheet_MobileHeaderTextColor";
  CustomVariables["ActionSheetTopBorderRadius"] = "--ui5wcr_ActionSheet_TopBorderRadius";
  CustomVariables["AnalyticalTableHeaderActiveTextColor"] = "--ui5wcr-AnalyticalTable-HeaderActiveTextColor";
  CustomVariables["AnalyticalTableHeaderBorderWidth"] = "--ui5wcr-AnalyticalTable-HeaderBorderWidth";
  CustomVariables["AnalyticalTableHeaderFontFamily"] = "--ui5wcr-AnalyticalTable-HeaderFontFamily";
  CustomVariables["AnalyticalTableOuterBorderBlock"] = "--ui5wcr-AnalyticalTable-OuterBorderBlock";
  CustomVariables["AnalyticalTableOuterBorderInline"] = "--ui5wcr-AnalyticalTable-OuterBorderInline";
  CustomVariables["ObjectPageSectionBorderTop"] = "--ui5wcr_ObjectPage_SectionBorderTop";
  CustomVariables["ObjectPageSectionTitleFontFamily"] = "--ui5wcr_ObjectPage_SectionTitleFontFamily";
  CustomVariables["ObjectPageSectionTitleHeight"] = "--ui5wcr_ObjectPage_SectionTitleHeight";
  CustomVariables["ObjectPageSectionTitleLineHeight"] = "--ui5wcr_ObjectPage_SectionTitleLineHeight";
  CustomVariables["ObjectPageSectionTitleTextDecoration"] = "--ui5wcr_ObjectPage_SectionTitleTextDecoration";
  CustomVariables["ObjectPageSubSectionBackgroundColor"] = "--ui5wcr_ObjectPage_SubSectionBackgroundColor";
  CustomVariables["ObjectPageSubSectionBorderRadius"] = "--ui5wcr_ObjectPage_SubSectionBorderRadius";
  CustomVariables["ObjectPageSubSectionPaddingBlockStart"] = "--ui5wcr_ObjectPage_PaddingBlockStart";
  CustomVariables["SplitterBarBorderFix"] = "--ui5wcr_Splitter_BarBorderHighContrastFix";
  CustomVariables["SplitterBarBorderFocus"] = "--ui5wcr_Splitter_BarBorderFocus";
  CustomVariables["SplitterBarBorderStyle"] = "--ui5wcr_Splitter_BarBorderStyle";
  CustomVariables["SplitterBarOutline"] = "--ui5wcr_Splitter_BarOutline";
  CustomVariables["SplitterContentBorderColor"] = "--ui5wcr_Splitter_ContentBorderColor";
  CustomVariables["SplitterIconColor"] = "--ui5wcr_Splitter_IconColor";
  CustomVariables["ToolbarFocusOutline"] = "--ui5wcr_Toolbar_FocusOutline";
  CustomVariables["ToolbarFocusShadow"] = "--ui5wcr_Toolbar_FocusShadow";
  return CustomVariables;
}({});
export const CustomThemingParameters = new Proxy(CustomVariables, {
  get: (target, prop) => `var(${target[prop]})`
});
