// @ui5/webcomponents-react/ssr/components/NumericSideIndicator/index.js
'use client';

import { clsx } from 'clsx';
import React, { forwardRef } from 'react';
import { createUseStyles } from 'react-jss';
import { ValueColor } from '../../enums';
import { Text } from '../Text';
import { NumericSideIndicatorStyles } from './NumericSideIndicator.jss';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
const useStyles = createUseStyles(NumericSideIndicatorStyles, {
  name: 'NumericSideIndicator'
});

/**
 * Holds a set of side indicator attributes used in the NumericHeader component.
 */
export const NumericSideIndicator = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    number,
    state,
    titleText,
    unit,
    className,
    ...rest
  } = props;
  const classes = useStyles();
  return /*#__PURE__*/_jsxs("div", {
    ref: ref,
    className: clsx(classes.numericSideIndicator, className),
    ...rest,
    children: [/*#__PURE__*/_jsx("span", {
      className: classes.title,
      children: titleText
    }), /*#__PURE__*/_jsxs("div", {
      className: clsx(classes.valueContainer, Reflect.get(classes, `state${state}`)),
      children: [/*#__PURE__*/_jsx(Text, {
        className: classes.number,
        children: number
      }), /*#__PURE__*/_jsx(Text, {
        className: classes.unit,
        children: unit
      })]
    })]
  });
});
NumericSideIndicator.displayName = 'NumericSideIndicator';
NumericSideIndicator.defaultProps = {
  state: ValueColor.None
};
